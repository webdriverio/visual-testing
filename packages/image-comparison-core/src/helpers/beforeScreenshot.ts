import logger from '@wdio/logger'
import hideScrollBars from '../clientSideScripts/hideScrollbars.js'
import setCustomCss from '../clientSideScripts/setCustomCss.js'
import { CUSTOM_CSS_ID } from './constants.js'
import { getAddressBarShadowPadding, getToolBarShadowPadding, waitFor } from './utils.js'
import getEnrichedInstanceData from '../methods/instanceData.js'
import type { BeforeScreenshotOptions, BeforeScreenshotResult } from './beforeScreenshot.interfaces.js'
import hideRemoveElements from '../clientSideScripts/hideRemoveElements.js'
import toggleTextTransparency from '../clientSideScripts/toggleTextTransparency.js'
import waitForFonts from '../clientSideScripts/waitForFonts.js'

const log = logger('@wdio/visual-service:beforeScreenshot')

/**
 * Methods that need to be executed before a screenshot will be taken
 */
export default async function beforeScreenshot(
    browserInstance: WebdriverIO.Browser,
    options: BeforeScreenshotOptions,
    addShadowPadding = false,
): Promise<BeforeScreenshotResult> {
    const { browserName, nativeWebScreenshot } = options.instanceData
    const { isAndroid, isIOS, isMobile } = browserInstance
    const {
        addressBarShadowPadding,
        disableBlinkingCursor,
        disableCSSAnimation,
        enableLayoutTesting,
        hideElements,
        noScrollBars,
        removeElements,
        toolBarShadowPadding,
        waitForFontsLoaded,
    } = options
    const addressBarPadding = getAddressBarShadowPadding({
        browserName,
        isAndroid,
        isIOS,
        isMobile,
        nativeWebScreenshot,
        addressBarShadowPadding,
        addShadowPadding,
    })
    const toolBarPadding = getToolBarShadowPadding({ isAndroid, isIOS, isMobile, browserName, toolBarShadowPadding, addShadowPadding })

    // Wait for the fonts to be loaded
    if (waitForFontsLoaded){
        try {
            await browserInstance.execute(waitForFonts)
        } catch (e) {
            log.debug('Waiting for fonts to load threw an error:', e)
        }
    }

    // Hide the scrollbars
    if (noScrollBars) {
        await browserInstance.execute(hideScrollBars, noScrollBars)
    }

    // Hide and or Remove elements
    if (hideElements.length > 0 || removeElements.length > 0) {
        try {
            await browserInstance.execute(hideRemoveElements, { hide: hideElements, remove: removeElements }, true)
        } catch (e) {
            log.warn(
                '\x1b[33m%s\x1b[0m',
                `
#####################################################################################
 WARNING:
 (One of) the elements that needed to be hidden or removed could not be found on the
 page and caused this error
 Error: ${e}
 We made sure the test didn't break.
#####################################################################################
`,
            )
        }
    }

    // Set some custom css
    if (disableCSSAnimation || disableBlinkingCursor || browserInstance.isMobile) {
        await browserInstance.execute(setCustomCss, { addressBarPadding, disableBlinkingCursor, disableCSSAnimation, id: CUSTOM_CSS_ID, toolBarPadding })
        // Wait at least 500 milliseconds to make sure the css is applied
        // Not every device is fast enough to apply the css faster
        await waitFor(500)
    }

    // Make all text transparent
    if (enableLayoutTesting){
        await browserInstance.execute(toggleTextTransparency, enableLayoutTesting)
        // Wait at least 500 milliseconds to make sure the css is applied
        // Not every device is fast enough to apply the css faster
        await waitFor(500)
    }

    // Get all the needed instance data
    const instanceOptions = {
        addressBarShadowPadding: options.addressBarShadowPadding,
        toolBarShadowPadding: options.toolBarShadowPadding,
        ...options.instanceData,
    }

    return getEnrichedInstanceData(browserInstance, instanceOptions, addShadowPadding)
}
