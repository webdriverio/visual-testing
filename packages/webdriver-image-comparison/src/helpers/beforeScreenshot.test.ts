import { describe, it, expect, vi } from 'vitest'
import beforeScreenshot from './beforeScreenshot.js'

describe('beforeScreenshot', () => {
    it('should be able to return the enriched instance data with default options', async () => {
        const MOCKED_EXECUTOR = vi.fn().mockReturnValue('')
        const options = {
            instanceData: {
                appName: 'appName',
                browserName: 'browserName',
                browserVersion: 'browserVersion',
                deviceName: 'deviceName',
                devicePixelRatio: 1,
                isAndroid: false,
                isIOS: false,
                isMobile: false,
                logName: 'logName',
                deviceRectangles: {
                    bottomBar: { y: 0, x: 0, width: 0, height: 0 },
                    homeBar: { x: 0, y:0, width: 0, height: 0 },
                    leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                    rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                    screenSize: { height: 1, width: 1 },
                    statusBar: { x: 0, y:0, width: 0, height: 0 },
                    statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
                    viewport: { y: 0, x: 0, width: 0, height: 0 },
                },
                name: 'name',
                nativeWebScreenshot: false,
                platformName: 'platformName',
                platformVersion: 'platformVersion',
                initialDevicePixelRatio: 1,
            },
            addressBarShadowPadding: 6,
            disableBlinkingCursor: false,
            disableCSSAnimation: false,
            enableLayoutTesting: false,
            noScrollBars: true,
            toolBarShadowPadding: 6,
            hideElements: [<HTMLElement>(<any>'<div></div>')],
            removeElements: [<HTMLElement>(<any>'<div></div>')],
            waitForFontsLoaded: true,
        }

        expect(await beforeScreenshot(MOCKED_EXECUTOR, options)).toMatchSnapshot()
    })

    it('should be able to return the enriched instance data with `addShadowPadding: true`', async () => {
        const MOCKED_EXECUTOR = vi.fn().mockReturnValue('')

        const options = {
            instanceData: {
                appName: 'appName',
                browserName: 'browserName',
                browserVersion: 'browserVersion',
                deviceName: 'deviceName',
                devicePixelRatio: 1,
                logName: 'logName',
                deviceRectangles: {
                    bottomBar: { y: 0, x: 0, width: 0, height: 0 },
                    homeBar: { x: 0, y:0, width: 0, height: 0 },
                    leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                    rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                    screenSize: { height: 1, width: 1 },
                    statusBar: { x: 0, y:0, width: 0, height: 0 },
                    statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
                    viewport: { y: 0, x: 0, width: 0, height: 0 },
                },
                isAndroid: false,
                isIOS: false,
                isMobile: false,
                name: 'name',
                nativeWebScreenshot: false,
                platformName: 'platformName',
                platformVersion: 'platformVersion',
                initialDevicePixelRatio: 1,
            },
            addressBarShadowPadding: 6,
            disableBlinkingCursor: true,
            disableCSSAnimation: true,
            enableLayoutTesting: false,
            noScrollBars: true,
            toolBarShadowPadding: 6,
            hideElements: [<HTMLElement>(<any>'<div></div>')],
            removeElements: [<HTMLElement>(<any>'<div></div>')],
            waitForFontsLoaded: true,
        }

        expect(await beforeScreenshot(MOCKED_EXECUTOR, options, true)).toMatchSnapshot()
    })
})
