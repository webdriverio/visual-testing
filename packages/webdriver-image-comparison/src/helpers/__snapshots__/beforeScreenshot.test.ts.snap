// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`beforeScreenshot > should be able to return the enriched instance data with \`addShadowPadding: true\` 1`] = `
{
  "addressBarShadowPadding": 0,
  "appName": "appName",
  "browserName": "browserName",
  "browserVersion": "browserVersion",
  "deviceName": "deviceName",
  "devicePixelRatio": 1,
  "deviceRectangles": {
    "bottomBar": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "homeBar": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "leftSidePadding": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "rightSidePadding": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "screenSize": {
      "height": 1,
      "width": 1,
    },
    "statusBar": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "statusBarAndAddressBar": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "viewport": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
  },
  "isAndroid": false,
  "isAndroidChromeDriverScreenshot": false,
  "isAndroidNativeWebScreenshot": false,
  "isIOS": false,
  "isMobile": false,
  "isTestInBrowser": true,
  "isTestInMobileBrowser": false,
  "logName": "logName",
  "name": "name",
  "nativeWebScreenshot": false,
  "platformName": "platformName",
  "platformVersion": "platformVersion",
  "toolBarShadowPadding": 0,
}
`;

exports[`beforeScreenshot > should be able to return the enriched instance data with default options 1`] = `
{
  "addressBarShadowPadding": 0,
  "appName": "appName",
  "browserName": "browserName",
  "browserVersion": "browserVersion",
  "deviceName": "deviceName",
  "devicePixelRatio": 1,
  "deviceRectangles": {
    "bottomBar": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "homeBar": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "leftSidePadding": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "rightSidePadding": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "screenSize": {
      "height": 1,
      "width": 1,
    },
    "statusBar": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "statusBarAndAddressBar": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "viewport": {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
  },
  "isAndroid": false,
  "isAndroidChromeDriverScreenshot": false,
  "isAndroidNativeWebScreenshot": false,
  "isIOS": false,
  "isMobile": false,
  "isTestInBrowser": true,
  "isTestInMobileBrowser": false,
  "logName": "logName",
  "name": "name",
  "nativeWebScreenshot": false,
  "platformName": "platformName",
  "platformVersion": "platformVersion",
  "toolBarShadowPadding": 0,
}
`;
